### ISSUE TITLE
Error related to stats::nls

### ISSUE THREAD

**Issue by ATpoint** [RWParsons/circacompare#3]
Dear Rex,

thank you for developing this package. I have the following error/problem to report.
Could you have a look?

best,
Alex

```r
library(circacompare)

y <- structure(list(time = c(1L, 1L, 1L, 1L, 5L, 5L, 9L, 9L, 13L, 
                             13L, 17L, 17L, 17L, 21L, 21L, 21L), 
                    value = c(6.46491702175632, 
                              6.37210528510888, 6.75505623236344, 6.4457897862926, 6.63766950190431, 
                              6.48725138475295, 6.40819847507183, 6.42253808100338, 6.37486222182972, 
                              6.51868394085349, 6.41506838906571, 6.40449437273951, 6.47273627195726, 
                              6.76905314588271, 6.59233676207294, 6.44481187866212)), 
                    class = "data.frame", row.names = c(NA, -16L))

set.seed(1)
circacompare::circa_single(x=y, col_time="time", col_outcome="value", return_figure=FALSE)
#> Error in stats::nls(formula = form, data = x, start = start_list(outcome = x$measure,  : 
#>   step factor 0.000488281 reduced below 'minFactor' of 0.000976562
#> Error in stats::nls(formula = form, data = x, start = start_list(outcome = x$measure,  : 
#>   step factor 0.000488281 reduced below 'minFactor' of 0.000976562
#> Error in stats::nls(formula = form, data = x, start = start_list(outcome = x$measure,  : 
#>   step factor 0.000488281 reduced below 'minFactor' of 0.000976562
#> $fit
#> Nonlinear regression model
#>   model: measure ~ k + (alpha) * cos((1/period) * time_r - (phi))
#>    data: x
#>       k   alpha     phi 
#> 6.48863 0.06986 6.26774 
#>  residual sum-of-squares: 0.1968
#> 
#> Number of iterations to convergence: 8 
#> Achieved convergence tolerance: 1.604e-08
#> 
#> $summary
#>         parameter       value
#> 1      rhythmic_p  0.13140327
#> 2           mesor  6.48862989
#> 3       amplitude  0.06985646
#> 4   phase_radians  6.26774340
#> 5 peak_time_hours 23.94101626
#> 6          period 24.00000000

packageVersion("circacompare")
#> [1] '0.1.1'

sessionInfo()
#> R version 4.2.1 (2022-06-23)
#> Platform: x86_64-pc-linux-gnu (64-bit)
#> Running under: Ubuntu 20.04.4 LTS
#> 
#> Matrix products: default
#> BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
#> LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3
#> 
#> locale:
#>  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
#>  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
#>  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
#>  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
#>  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
#> [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#> [1] circacompare_0.1.1
#> 
#> loaded via a namespace (and not attached):
#>  [1] compiler_4.2.1   pillar_1.8.0     highr_0.9        tools_4.2.1     
#>  [5] digest_0.6.29    evaluate_0.16    lifecycle_1.0.1  tibble_3.1.8    
#>  [9] gtable_0.3.0     pkgconfig_2.0.3  rlang_1.0.4      reprex_2.0.1    
#> [13] DBI_1.1.3        cli_3.3.0        rstudioapi_0.13  yaml_2.3.5      
#> [17] xfun_0.32        fastmap_1.1.0    withr_2.5.0      stringr_1.4.0   
#> [21] dplyr_1.0.9      knitr_1.39       generics_0.1.3   fs_1.5.2        
#> [25] vctrs_0.4.1      grid_4.2.1       tidyselect_1.1.2 glue_1.6.2      
#> [29] R6_2.5.1         fansi_1.0.3      rmarkdown_2.14   purrr_0.3.4     
#> [33] ggplot2_3.3.6    magrittr_2.0.3   scales_1.2.0     htmltools_0.5.3 
#> [37] assertthat_0.2.1 colorspace_2.0-3 utf8_1.2.2       stringi_1.7.8   
#> [41] munsell_0.5.0
```

**Comment by RWParsons** [comment]
Hi Alex,

This is a bit of an unavoidable feature of using nls() to fit the models. Sometimes they will converge elsewhere or have problems fitting. The "solution" implemented in circacompare to address this is to repeat the nls fitting procedure until it succeeds and the parameter estimates are sensible (positive amplitude, acrophase within 0 and 2*pi, etc). The errors that appear in the console are appearing when nls attempts to fit but fails, this procedure is reattempted and that is the resulting fit that's (eventually) returned.

I hope this clears up what is happening. I have chosen not to suppress these messages to the user but, if you think it'd be best, I could add an argument that would let you control whether they're hidden or not? 

Thanks,
Rex

**Comment by ATpoint** [comment]
Thank you for the response. I think an option to suppress would be desirable, since usual methods like `try(x, silent=TRUE)` and `suppressMessages()` don't do the trick.

One more question if I may (possible feature request)

=> I am in a situation where our data suffer from some outliers and nested batch effects. There was no way to correct for that by standard methods (like RUVseq), removing outliers reduced power and created quite unequal sample size between groups. What did the trick nicely while calling rhythmic genes with limma (cosinor-like, as in limorhyde) was to use the arrayWeights function from limma to downweight outlier. Therefore the question whether something like this is basically possible in circacompare/nls, so accepting a set of user-provided weights?

**Comment by RWParsons** [comment]
Hi Alex,

No worries. I have updated the development version so that it now includes a "suppress_all" argument. If you set this to TRUE, it'll hide all messages including the ones you show in your example above. Install using:
`remotes::install_github("RWParsons/circacompare")`

I have also just investigated how to pass weights to nls and nlme but unfortunately it's a bit more tricky than I anticipated and I don't have time to get it in the near future. I'm going to close this issue but if you want to open a new one for the weights, then I'll try to get to it later.
